ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -376,81 +477,169 @@ pub(crate) enum Step {
+/// It is different from the validator trait since context outputs a `DZKPUpgradedContext`
+/// that is tied to a `DZKPBatch` rather than an `accumulator`.
+/// Function signature of `validate` is also different, it does not downgrade shares anymore
 #[async_trait]
-pub trait DZKPValidator<DF: DZKPBaseField, B: DZKPUpgradableContext<DF>> {
-    fn context(&self) -> B::UpgradedContext;
-    async fn validate(self) -> Result<(), Error>;
+pub trait DZKPValidator<B: UpgradableContext> {
+    fn context(&self) -> B::DZKPUpgradedContext;
+
+    /// Allows to validate the current `DZKPBatch` and empties it. The associated context is then
+    /// considered safe until another multiplication is performed and thus new values are added
+    /// to `DZKPBatch`.
+    /// Is generic over `DZKPBaseFields`. Please specify a sufficiently large field for the current `DZKPBatch`.
+    async fn validate<DF: DZKPBaseField>(&self) -> Result<(), Error>;
+
+    /// `is_safe` checks that there are no remaining `UnverifiedValues` within the associated `DZKPBatch`
+    ///
+    /// ## Errors
+    /// Errors when there are `UnverifiedValues` left.
+    fn is_safe(&self) -> Result<(), Error>;
+
+    /// `get_chunk_size` returns the chunk size of the validator
+    fn get_chunk_size(&self) -> Option<usize>;
 }
 
-pub struct SemiHonestDZKPValidator<'a, DF> {
-    context: SemiHonestDZKPUpgraded<'a>,
-    _f: PhantomData<DF>,
+#[allow(dead_code)]
+/// `validated_seq_join` is a validated `seq_join`. It splits the input stream into `chunks` where
+/// the `chunk_size` is defined within `validator`. Each `chunk` is a vector that is independently
+/// verified using `validator.validate()`, which uses DZKPs. Once the validation fails,
+/// the output stream will return that the stream is done.
+///
+// Ideally it would be part of the DZKPValidator trait but trait functions don't like impl generics return types
+fn validated_seq_join<'st, S, F, O, DF, B, V>(
+    validator: &'st V,
+    source: S,
+) -> impl Stream<Item = Result<O, Error>> + 'st
+where
+    S: Stream<Item = F> + Send + 'st,
+    F: Future<Output = O> + Send + 'st,
+    O: Send + Sync + Clone + 'static,
+    DF: DZKPBaseField,
+    B: UpgradableContext,
+    V: DZKPValidator<B> + Send + Sync,
+{
+    // chunk_size is undefined in the semi-honest setting, set it to 10, ideally it would be 1
+    // but there is some overhead
+    let chunk_size = validator.get_chunk_size().unwrap_or(10usize);
+    seq_join::<'st, S, F, O>(validator.context().active_work(), source)
+        .chunks(chunk_size)
+        .then(move |x| async move {
+            let valid = validator.validate::<DF>().await.is_ok();
+            iter(x).map(move |x| {
+                if valid {
+                    Ok(x)
+                } else {
+                    Err(Error::DZKPValidationFailed)
+                }
+            })
+        })
+        .flatten()
+        .take_while(|x| future::ready(x.is_ok()))
```

Comment:
I like that your code is very concise. I had issues to propagate the error from a chunk to every single element since error does not implement clone. I also did not want to use an Arc or here. How does try_flatten_iters work? Alex suggested to terminate the stream when we hit an error. I think he was concerned that you could ignore the error and just move on with the next element that is not an error which would not be safe against malicious helper parties.